README.md
# varuint-bitcoin

encode/decode number
as [bitcoin variable length integer](https://en.bitcoin.it/wiki/Protocol_documentation#Variable_length_integer)

| value | storage length (bytes) |
|:------:|:--------------:|
| [0, 0xfd) | 1 |
| [0xfd, 0xffff] | 3 |
| [0x010000, 0xffffffff] | 5 |
| [0x0100000000, 0x1fffffffffffff] | 9 |

## Installation

    dart pub add dart_varuint_bitcoin

## Example

```dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_varuint_bitcoin/dart_varuint_bitcoin.dart' as varuint;

void main() {
  Uint8List buffer = Uint8List(2);
  final encoded = varuint.encode(0xfc, buffer, 1);
  print(hex.encode(encoded.buffer.buffer.asUint8List()));
  // => 00fx

  Uint8List buffer2 = Uint8List.fromList([0x00, 0xfc]);
  final decoded = varuint.decode(buffer2, 1);
  print(decoded.output.toRadixString(16));
  // => fc
}
```

## License

MIT

CHANGELOG.md
## 1.0.3

- update ByteData types to Uint8List

## 1.0.2

- add typing

## 1.0.1

- add api docs

## 1.0.0

- Initial version.


test/fixtures.dart
enum FixtureEnum { valid, invalid }

class Fixture {
  String? msg;
  int? dec;
  String? hex;

  Fixture({this.hex, this.dec, this.msg});
}

Map<FixtureEnum, List<Fixture>> fixtures = {
  FixtureEnum.valid: [
    Fixture(dec: 0, hex: "00"),
    Fixture(dec: 1, hex: "01"),
    Fixture(dec: 252, hex: "fc"),
    Fixture(dec: 253, hex: "fdfd00"),
    Fixture(dec: 254, hex: "fdfe00"),
    Fixture(dec: 255, hex: "fdff00"),
    Fixture(dec: 65534, hex: "fdfeff"),
    Fixture(dec: 65535, hex: "fdffff"),
    Fixture(dec: 65536, hex: "fe00000100"),
    Fixture(dec: 65537, hex: "fe01000100"),
    Fixture(dec: 4294967295, hex: "feffffffff"),
    Fixture(dec: 4294967296, hex: "ff0000000001000000"),
    Fixture(dec: 4294967297, hex: "ff0100000001000000"),
    Fixture(dec: 9007199254740991, hex: "ffffffffffffff1f00")
  ],
  FixtureEnum.invalid: [
    Fixture(dec: -1, msg: 'RangeError: value out of range'),
    Fixture(
        dec: 9007199254740992,
        hex: "ffffffffff00002000",
        msg: 'RangeError: value out of range')
  ]
};


test/dart_varuint_bitcoin_test.dart
import 'dart:typed_data';

import 'package:convert/convert.dart';
import 'package:dart_varuint_bitcoin/dart_varuint_bitcoin.dart' as varuint;
import 'package:test/test.dart';

import 'fixtures.dart';

void main() {
  if (fixtures.containsKey(FixtureEnum.valid)) {
    for (Fixture fixture in fixtures[FixtureEnum.valid]!) {
      test('Valid encode #${fixture.hex} (${fixture.dec!.toString()})', () {
        final encoded = varuint.encode(fixture.dec!);
        expect(hex.encode(encoded.buffer.buffer.asUint8List()),
            equals(fixture.hex));
        expect(encoded.bytes, equals(fixture.hex!.length / 2));
      });
      test('Valid decode #${fixture.hex} (${fixture.dec!.toString()})', () {
        final decoded =
            varuint.decode(Uint8List.fromList(hex.decode(fixture.hex!)));
        expect(decoded.output, equals(fixture.dec));
        expect(decoded.bytes, equals(fixture.hex!.length / 2));
      });
      test('Valid encode #${fixture.hex} (${fixture.dec!.toString()})', () {
        expect(varuint.encodingLength(fixture.dec!),
            equals(fixture.hex!.length / 2));
      });
    }
  }
  if (fixtures.containsKey(FixtureEnum.invalid)) {
    for (Fixture fixture in fixtures[FixtureEnum.invalid]!) {
      test('Invalid for #${fixture.hex} (${fixture.dec!.toString()})', () {
        try {
          varuint.encode(fixture.dec!);
        } catch (e) {
          expect(e, isRangeError);
          expect((e as RangeError).toString(), equals(fixture.msg));
          return;
        }

        throw Exception('Should be throw');
      });
    }
  }

  test('write to buffer with offset', () {
    Uint8List buffer = Uint8List(2);
    final encoded = varuint.encode(0xfc, buffer, 1);
    expect(hex.encode(encoded.buffer.buffer.asUint8List()), equals('00fc'));
    expect(encoded.bytes, 1);
  });

  test('should be a buffer', () {
    Uint8List buffer = Uint8List.fromList([0x00, 0xfc]);
    final decoded = varuint.decode(buffer, 1);
    expect(decoded.output, equals(0xfc));
    expect(decoded.bytes, 1);
  });
}


lib/src/dart_varuint_bitcoin_base.dart
import 'dart:typed_data';

var _maxSafeInteger = 9007199254740991;

void _checkUInt53(int n) {
  if (n < 0 || n > _maxSafeInteger) {
    throw RangeError('value out of range');
  }
}

class _EncodeOutputModel {
  final int bytes;

  final Uint8List buffer;

  _EncodeOutputModel(this.bytes, this.buffer);

  @override
  String toString() {
    return buffer.toString();
  }
}

class _DecodeOutputModel {
  final int bytes;

  final int output;

  _DecodeOutputModel(this.bytes, this.output);

  @override
  String toString() {
    return output.toString();
  }
}

/// Encode varuint
_EncodeOutputModel encode(int number, [Uint8List? buffer, int offset = 0]) {
  _checkUInt53(number);

  int bytes = 0;

  buffer ??= Uint8List(encodingLength(number));

  ByteData bufferByteData = buffer.buffer.asByteData();

  // 8 bit
  if (number < 0xfd) {
    bufferByteData.setUint8(offset, number);
    bytes = 1;

    // 16 bit
  } else if (number <= 0xffff) {
    bufferByteData.setUint8(offset, 0xfd);
    bufferByteData.setUint16(offset + 1, number, Endian.little);
    bytes = 3;

    // 32 bit
  } else if (number <= 0xffffffff) {
    bufferByteData.setUint8(offset, 0xfe);
    bufferByteData.setUint32(offset + 1, number, Endian.little);
    bytes = 5;

    // 64 bit
  } else {
    bufferByteData.setUint8(offset, 0xff);
    bufferByteData.setUint64(offset + 1, number, Endian.little);
    bytes = 9;
  }

  return _EncodeOutputModel(bytes, buffer);
}

/// Decode varuint
_DecodeOutputModel decode(Uint8List buffer, [int offset = 0]) {
  int bytes = 0, output;

  ByteData bufferByteData = buffer.buffer.asByteData();
  int first = bufferByteData.getUint8(offset);

  // 8 bit
  if (first < 0xfd) {
    bytes = 1;
    output = first;
    // 16 bit
  } else if (first == 0xfd) {
    bytes = 3;
    output = bufferByteData.getUint16(offset + 1, Endian.little);

    // 32 bit
  } else if (first == 0xfe) {
    bytes = 5;
    output = bufferByteData.getUint32(offset + 1, Endian.little);

    // 64 bit
  } else {
    bytes = 9;
    output = bufferByteData.getUint64(offset + 1, Endian.little);
  }

  return _DecodeOutputModel(bytes, output);
}

/// Get encoding length of varuint
int encodingLength(int number) {
  _checkUInt53(number);

  return (number < 0xfd
      ? 1
      : number <= 0xffff
          ? 3
          : number <= 0xffffffff
              ? 5
              : 9);
}
